src/bitcoin.cpp

```diff
@@ -0,0 +1,204 @@
+        return EXIT_FAILURE;
+    }
+    return EXIT_SUCCESS;
+}
+
+CommandLine ParseCommandLine(int argc, char* argv[])
+{
+    CommandLine cmd;
+    cmd.args.reserve(argc);
+    for (int i = 1; i < argc; ++i) {
+        std::string_view arg = argv[i];
+        if (!cmd.command.empty()) {
+            cmd.args.emplace_back(argv[i]);
+        } else if (arg == "-m" || arg == "--multiprocess") {
+            cmd.use_multiprocess = true;
+        } else if (arg == "-M" || arg == "--monolithic") {
+            cmd.use_multiprocess = false;
+        } else if (arg == "-v" || arg == "--version") {
+            cmd.show_version = true;
+        } else if (arg == "-h" || arg == "--help" || arg == "help") {
+            cmd.show_help = true;
+        } else if (cmd.show_help && (arg == "-a" || arg == "--all")) {
+            cmd.show_help_all = true;
+        } else if (arg.starts_with("-")) {
+            throw std::runtime_error(strprintf("Unknown option: %s", arg));
+        } else if (!arg.empty()) {
+            cmd.command = arg;
+        }
+    }
+    return cmd;
+}
+
+//! Execute the specified bitcoind, bitcoin-qt or other command line in `args`
+//! using src, bin and libexec directory paths relative to this executable, where
+//! the path to this executable is specified in `wrapper_argv0`.
+//!
+//! @param args Command line arguments to execute, where first argument should
+//!             be a relative path to a bitcoind, bitcoin-qt or other executable
+//!             that will be located on the PATH or relative to wrapper_argv0.
+//!
+//! @param wrapper_argv0 String containing first command line argument passed to
+//!                      main() to run the current executable. This is used to
+//!                      help determine the path to the current executable and
+//!                      how to look for new executables.
+//
+//! @note This function doesn't currently print anything but can be debugged
+//! from the command line using strace or dtrace like:
+//!
+//!     strace -e trace=execve -s 10000 build/bin/bitcoin ...
+//!     dtrace -n 'proc:::exec-success  /pid == $target/ { trace(curpsinfo->pr_psargs); }' -c ...
+static void ExecCommand(const std::vector<const char*>& args, std::string_view wrapper_argv0)
+{
+    // Construct argument string for execvp
+    std::vector<const char*> exec_args{args};
+    exec_args.emplace_back(nullptr);
+
+    // Try to call ExecVp with given exe path.
+    auto try_exec = [&](fs::path exe_path, bool allow_notfound = true) {
+        std::string exe_path_str{fs::PathToString(exe_path)};
+        exec_args[0] = exe_path_str.c_str();
+        if (util::ExecVp(exec_args[0], (char*const*)exec_args.data()) == -1) {
+            if (allow_notfound && errno == ENOENT) return false;
+            throw std::system_error(errno, std::system_category(), strprintf("execvp failed to execute '%s'", exec_args[0]));
+        }
+        throw std::runtime_error("execvp returned unexpectedly");
```

Comment:
This can be written like:

```cpp
util::ExecVp(exec_args[0], (char*const*)exec_args.data()); // If it returns then an error has occurred.

if (allow_notfound && errno == ENOENT) return false;
throw std::system_error(errno, std::system_category(), strprintf("execvp failed to execute '%s'", exec_args[0]));
```
it is less indentation and less code - no need for the "this will never be reached" `throw`.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/31375#discussion_r2099737369

execvp is documented to return -1 if there is an error (https://linux.die.net/man/3/execvp and https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/execvp-wexecvp) so if it returns something else I think it is better to abort and not just keep going.